game/world/objects/npc.cpp

```diff
@@ -3447,26 +3457,35 @@ bool Npc::tickCast(uint64_t dt) {
   if(bodyStateMasked()!=BS_CASTING)
     return true;
 
-  auto& spl = owner.script().spellDesc(active->spellId());
-  int32_t castLvl = int(castLevel)-int(CS_Invest_0);
   if(owner.tickCount()<castNextTime)
     return true;
 
-  if(castLvl<=15)
-    castLevel = CastState(castLevel+1);
+  ++manaInvested;
```

Comment:
In G2 the amout of mana needed to reach a new cast level is constant per spell. But G1 script has lines like 
```
	if 	(manaInvested ==(SPL_SENDCAST_FIRESTORM*3)/10)
	||	(manaInvested ==(SPL_SENDCAST_FIRESTORM*5)/10)
	||	(manaInvested ==(SPL_SENDCAST_FIRESTORM*8)/10)
	{	return SPL_NEXTLEVEL;		};
```
where mana needed for next level can be arbitrary.

Reply:
In G1 player's mana is reduced in 1-steps and script comments stated `manaInvested` is increased by 1 per `Spell_Logic` call by engine. There's no mana attribute handling in G1 scripts so I added `changeAttribute(ATR_MANA,-1,false);`

For G2 i tried to get the needed mana somehow from values like `const int STEP_ChargeFireball = 40;` but after that was unsatisfying I added a print command to scripts to check the value in vanilla. And indeed in G1 and G2 `manaInvested` is a `0,1,2,... ` sequence. But here mana attribute changes are handled by script.  

This handling in G2  inherits the vanilla bug that if e.g. players' mana is 70 and mana per level 50, spell is charged to level 2 and player's mana is set to 0. So one more level than mana would allow for if player's start mana is in spell's `[min,max)` interval.

Reply:
In G2-notr, there is `Spell_Logic_SuckEnergy` only spell that uses `SPL_RECEIVEINVEST` code, do not decrement mana, seemingly relay on engine?

Reply:
>Should it be 1 per SPL_RECEIVEINVEST ?

I retested, `
 SPL_RECEIVEINVEST;
  SPL_STATUS_CANINVEST_NO_MANADEC;
  SPL_NEXTLEVEL;
`  all give +1.

`SPL_DONTINVEST` is only used  in G2 to hold a maximal charged spell. `Spell_Logic` is not called anymore and spell is sent on player action. 

>In G2-notr, there is Spell_Logic_SuckEnergy only spell that uses SPL_RECEIVEINVEST code, do not decrement mana, seemingly relay on engine?

I thought this state was only in use by G1, missed that one. Mana decrement is handled in `Spell_Cast_SuckEnergy`.

In G1 there's line `if (manaInvested == 0) { return SPL_NEXTLEVEL;	};` in firerain file. Comment hints it's needed to play invest fire effect. Maybe it's something similar for `SuckEnergy`.